tools/build_updated_packages.rb

```diff
@@ -0,0 +1,177 @@
+boolean_properties = `sed -n '/^\ \ boolean_property/,/^$/p' lib/package.rb | sed 's/^\ \ boolean_property//' | tr -d '\n' | sort -u`.chomp.delete(',').delete(':').split.join(' ')
+boolean_property(boolean_properties.split)
+
+# Get all other properties from buildsystems files and package.rb
+buildsystem_properties = `sed -n '/^\ \ property/,/^$/p' lib/buildsystems/* | sed 's/^\ \ property//' | tr -d '\n' | sort -u`.chomp.delete(',').delete(':').split.join(' ')
+package_properties = `sed -n '/^\ \ property/,/^$/p' lib/package.rb | sed 's/^\ \ property//' | tr -d '\n' | sort -u`.chomp.delete(',').delete(':').split.join(' ')
+properties = "#{package_properties} #{buildsystem_properties}"
+property(properties.split)
+
+puts 'Checking for pip package version updates...'.orange
+system '../tools/update_python_pip_packages.rb', chdir: 'packages'
+puts 'Checking for ruby gem package version updates...'.orange
+system '../tools/update_ruby_gem_packages.rb', chdir: 'packages'
+changed_files = `git diff HEAD --name-only`.chomp.split
+updated_packages = changed_files.select { |c| c.include?('packages/') }
+
+if updated_packages.empty?
+  puts 'No packages need to be updated.'.orange
+else
+  puts 'These changed packages will be checked to see if they need updated binaries:'.orange
+  updated_packages.each { |p| puts p.sub('packages/', '').sub('.rb', '').to_s.lightblue }
+end
+
+base_url = 'https://gitlab.com/api/v4/projects/26210301/packages'
+
+updated_packages.each do |pkg|
+  name = pkg.sub('packages/', '').sub('.rb', '')
+
+  puts "Evaluating #{name} package...".orange
+  # rubocop:disable Security/Eval
+  eval(`sed -e '/^require/d' -e '/^\ \ depends_on/d' -e '/^class/d' -e '/^end/d' #{pkg}`.chomp)
+  # rubocop:enable Security/Eval
+  boolean_property(boolean_properties.split)
+  property(properties.split)
+
+  # Don't check if we need new binaries if the package doesn't already
+  # have binaries for this architecture and no_compile_needed is set.
+  if !system("grep -q binary_sha256 #{pkg}") && no_compile_needed?
+    puts "#{name.capitalize} #{@version} has no binaries and may not need them.".lightgreen
+    next pkg
+  else
+    build = compatibility == 'all' ? %w[x86_64 aarch64 armv7l i686] : compatibility.split
+    build.each do |arch|
+      arch_specific_url = "#{base_url}/generic/#{name}/#{@version}_#{arch}/#{name}-#{@version}-chromeos-#{arch}.tar.zst"
+      if `curl -sI #{arch_specific_url}`.lines.first.split[1] == '200'
+        build.delete(arch)
+        next
+      end
+    end
+    if build.empty?
+      puts "No builds are needed for #{name} #{@version}.".lightgreen
+      next
+    else
+      puts "#{name.capitalize} #{@version} needs builds uploaded for: #{build.join(' ')}".lightblue
+    end
+    if build.include?(ARCH) && agree_default_yes("\nWould you like to build #{name} #{@version}")
+      system "yes | crew build -f #{pkg}"
+      system "crew upload #{name}"
+    end
+  end
+end
```

Comment:
I have to admit, this is great work and I'm sure it took some time.

Reply:
Yeah, this is a case of `burying the lede` in journalism terms. This entire PR is mostly in service of making this script work...